x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/hooks/use_step_define_form.ts

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { useEffect } from 'react';
+
+import { getPreviewRequestBody } from '../../../../../common';
+
+import { getDefaultStepDefineState } from '../common';
+
+import { StepDefineFormProps } from '../step_define_form';
+
+import { useAdvancedPivotEditor } from './use_advanced_pivot_editor';
+import { useAdvancedSourceEditor } from './use_advanced_source_editor';
+import { usePivotConfig } from './use_pivot_config';
+import { useSearchBar } from './use_search_bar';
+
+export type StepDefineFormHook = ReturnType<typeof useStepDefineForm>;
+
+export const useStepDefineForm = ({ overrides, onChange, searchItems }: StepDefineFormProps) => {
+  const defaults = { ...getDefaultStepDefineState(searchItems), ...overrides };
+  const { indexPattern } = searchItems;
+
+  const searchBar = useSearchBar(defaults, indexPattern);
+  const pivotConfig = usePivotConfig(defaults, indexPattern);
+
+  const previewRequest = getPreviewRequestBody(
+    indexPattern.title,
+    searchBar.state.pivotQuery,
+    pivotConfig.state.pivotGroupByArr,
+    pivotConfig.state.pivotAggsArr
+  );
+
+  // pivot config hook
+  const stringifiedPivotConfig = JSON.stringify(previewRequest.pivot, null, 2);
+  const advancedPivotEditor = useAdvancedPivotEditor(defaults, stringifiedPivotConfig);
+
+  // source config hook
+  const stringifiedSourceConfig = JSON.stringify(previewRequest.source.query, null, 2);
+  const advancedSourceEditor = useAdvancedSourceEditor(defaults, stringifiedSourceConfig);
```

Comment:
maybe the hooks itself should be responsible for stringifying configs?

Reply:
Moved it to the hook itself in 3b773ef.